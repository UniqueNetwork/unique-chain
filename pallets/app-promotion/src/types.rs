use codec::EncodeLike;
use frame_support::{
	traits::LockableCurrency, WeakBoundedVec, Parameter, dispatch::DispatchResult, ensure,
};

use pallet_balances::{BalanceLock, Config as BalancesConfig, Pallet as PalletBalances};
use pallet_common::CollectionHandle;
use pallet_unique::{Event as UniqueEvent, Error as UniqueError};
use sp_runtime::DispatchError;
use up_data_structs::{CollectionId, SponsorshipState};
use sp_std::borrow::ToOwned;
use pallet_evm_contract_helpers::{Pallet as EvmHelpersPallet, Config as EvmHelpersConfig, Sponsoring};

pub trait ExtendedLockableCurrency<AccountId: Parameter>: LockableCurrency<AccountId> {
	fn locks<KArg>(who: KArg) -> WeakBoundedVec<BalanceLock<Self::Balance>, Self::MaxLocks>
	where
		KArg: EncodeLike<AccountId>;
}

impl<T: BalancesConfig<I>, I: 'static> ExtendedLockableCurrency<T::AccountId>
	for PalletBalances<T, I>
{
	fn locks<KArg>(who: KArg) -> WeakBoundedVec<BalanceLock<Self::Balance>, Self::MaxLocks>
	where
		KArg: EncodeLike<T::AccountId>,
	{
		Self::locks(who)
	}
}

pub trait CollectionHandler {
	type CollectionId;
	type AccountId;

	fn set_sponsor(
		sponsor_id: Self::AccountId,
		collection_id: Self::CollectionId,
	) -> DispatchResult;

	fn remove_collection_sponsor(collection_id: Self::CollectionId) -> DispatchResult;

	fn get_sponsor(
		collection_id: Self::CollectionId,
	) -> Result<Option<Self::AccountId>, DispatchError>;
}

impl<T: pallet_unique::Config> CollectionHandler for pallet_unique::Pallet<T> {
	type CollectionId = CollectionId;

	type AccountId = T::AccountId;

	fn set_sponsor(
		sponsor_id: Self::AccountId,
		collection_id: Self::CollectionId,
	) -> DispatchResult {
		let mut target_collection = <CollectionHandle<T>>::try_get(collection_id)?;
		target_collection.check_is_internal()?;
		target_collection.set_sponsor(sponsor_id.clone())?;

		Self::deposit_event(UniqueEvent::<T>::CollectionSponsorSet(
			collection_id,
			sponsor_id.clone(),
		));

		ensure!(
			target_collection.confirm_sponsorship(&sponsor_id)?,
			UniqueError::<T>::ConfirmUnsetSponsorFail
		);

		Self::deposit_event(UniqueEvent::<T>::SponsorshipConfirmed(
			collection_id,
			sponsor_id,
		));

		target_collection.save()
	}

	fn remove_collection_sponsor(collection_id: Self::CollectionId) -> DispatchResult {
		let mut target_collection = <CollectionHandle<T>>::try_get(collection_id)?;
		target_collection.check_is_internal()?;
		target_collection.sponsorship = SponsorshipState::Disabled;

		Self::deposit_event(UniqueEvent::<T>::CollectionSponsorRemoved(collection_id));

		target_collection.save()
	}

	fn get_sponsor(
		collection_id: Self::CollectionId,
	) -> Result<Option<Self::AccountId>, DispatchError> {
		Ok(<CollectionHandle<T>>::try_get(collection_id)?
			.sponsorship
			.pending_sponsor()
			.map(|acc| acc.to_owned()))
	}
}

pub trait ContractHandler {
	type ContractId;
	type AccountId;

	fn set_sponsor(sponsor_id: Self::AccountId, contract_id: Self::ContractId) -> DispatchResult;

	fn remove_contract_sponsor(collection_id: Self::ContractId) -> DispatchResult;

	fn get_sponsor(contract_id: Self::ContractId)
		-> Result<Option<Self::AccountId>, DispatchError>;
}

impl<T: EvmHelpersConfig> ContractHandler for EvmHelpersPallet<T> {
	type ContractId = sp_core::H160;

	type AccountId = T::CrossAccountId;

	fn set_sponsor(sponsor_id: Self::AccountId, contract_id: Self::ContractId) -> DispatchResult {
		Sponsoring::<T>::insert(
			contract_id,
			SponsorshipState::<T::CrossAccountId>::Confirmed(sponsor_id),
		);
		Ok(())
	}

	fn remove_contract_sponsor(contract_id: Self::ContractId) -> DispatchResult {
		Sponsoring::<T>::remove(contract_id);
		Ok(())
	}

	fn get_sponsor(
		contract_id: Self::ContractId,
	) -> Result<Option<Self::AccountId>, DispatchError> {
		Ok(Self::get_sponsor(contract_id))
	}
}
