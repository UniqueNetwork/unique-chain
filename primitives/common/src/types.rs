// Copyright 2019-2022 Unique Network (Gibraltar) Ltd.
// This file is part of Unique Network.

// Unique Network is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Unique Network is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Unique Network. If not, see <http://www.gnu.org/licenses/>.

use sp_runtime::{
	generic,
	traits::{IdentifyAccount, Verify},
	MultiSignature,
};

/// Opaque types. These are used by the CLI to instantiate machinery that don't need to know
/// the specifics of the runtime. They can then be made to be agnostic over specific formats
/// of data like extrinsics, allowing for them to continue syncing the network through upgrades
/// to even the core data structures.
pub mod opaque {
	pub use sp_runtime::{generic, traits::BlakeTwo256, OpaqueExtrinsic as UncheckedExtrinsic};

	pub use super::{AccountId, AuraId, Balance, BlockNumber, Hash, Signature};

	#[derive(Debug, Clone)]
	pub enum RuntimeId {
		Unique,
		Quartz,
		Opal,
		Unknown(sp_std::vec::Vec<u8>),
	}

	pub type Header = generic::Header<BlockNumber, BlakeTwo256>;

	pub type Block = generic::Block<Header, UncheckedExtrinsic>;

	pub trait RuntimeInstance {
		type CrossAccountId: pallet_evm::account::CrossAccountId<sp_runtime::AccountId32>
			+ Send
			+ Sync
			+ 'static;
	}
}

pub type SessionHandlers = ();

/// An index to a block.
pub type BlockNumber = u32;

/// Alias to 512-bit hash when used in the context of a transaction signature on the chain.
pub type Signature = MultiSignature;

/// Some way of identifying an account on the chain. We intentionally make it equivalent
/// to the public key of our transaction signing scheme.
pub type AccountId = <<Signature as Verify>::Signer as IdentifyAccount>::AccountId;

/// The type for looking up accounts. We don't expect more than 4 billion of them, but you
/// never know...
pub type AccountIndex = u32;

/// Balance of an account.
pub type Balance = u128;

/// Index of a transaction in the chain.
pub type Nonce = u32;

/// A hash of some data used by the chain.
pub type Hash = sp_core::H256;

/// Digest item type.
pub type DigestItem = generic::DigestItem;

pub use sp_consensus_aura::sr25519::AuthorityId as AuraId;
