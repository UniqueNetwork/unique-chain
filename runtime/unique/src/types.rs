use frame_support::traits::Currency;
use pallet_evm;
use sp_runtime;

use unique_runtime_common::types::AccountId;

use super::constructor::{Balances, Runtime};

pub(crate) type CrossAccountId = pallet_evm::account::BasicCrossAccountId<Runtime>;

pub(crate) type NegativeImbalance = <Balances as Currency<AccountId>>::NegativeImbalance;

/// Digest item type.
pub type DigestItem = sp_runtime::generic::DigestItem;

