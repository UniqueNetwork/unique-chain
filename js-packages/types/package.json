{
    "author": "",
    "license": "SEE LICENSE IN ../../../LICENSE",
    "description": "Interfaces for interacting with contracts and contract ABIs",
    "engines": {
        "node": ">=16"
    },
    "name": "@unique-nft/opal-testnet-types",
    "type": "module",
    "version": "1.0.0",
    "main": "index.js",
    "devDependencies": {
        "@polkadot/typegen": "^10.10.1"
    }
}
