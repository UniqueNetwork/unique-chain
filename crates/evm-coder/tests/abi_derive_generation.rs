use evm_coder_procedural::AbiCoder;
use evm_coder::types::bytes;

#[test]
fn empty_struct() {
	let t = trybuild::TestCases::new();
	t.compile_fail("tests/build_failed/abi_derive_generation.rs");
}

#[derive(AbiCoder, PartialEq, Debug)]
struct TypeStruct1SimpleParam {
	_a: u8,
}

#[derive(AbiCoder, PartialEq, Debug)]
struct TypeStruct1DynamicParam {
	_a: String,
}

#[derive(AbiCoder, PartialEq, Debug)]
struct TypeStruct2SimpleParam {
	_a: u8,
	_b: u32,
}

#[derive(AbiCoder, PartialEq, Debug)]
struct TypeStruct2DynamicParam {
	_a: String,
	_b: bytes,
}

#[derive(AbiCoder, PartialEq, Debug)]
struct TypeStruct2MixedParam {
	_a: u8,
	_b: bytes,
}

#[derive(AbiCoder, PartialEq, Debug)]
struct TypeStruct1DerivedSimpleParam {
	_a: TypeStruct1SimpleParam,
}

#[derive(AbiCoder, PartialEq, Debug)]
struct TypeStruct2DerivedSimpleParam {
	_a: TypeStruct1SimpleParam,
	_b: TypeStruct2SimpleParam,
}

#[derive(AbiCoder, PartialEq, Debug)]
struct TypeStruct1DerivedDynamicParam {
	_a: TypeStruct1DynamicParam,
}

#[derive(AbiCoder, PartialEq, Debug)]
struct TypeStruct2DerivedDynamicParam {
	_a: TypeStruct1DynamicParam,
	_b: TypeStruct2DynamicParam,
}

#[derive(AbiCoder, PartialEq, Debug)]
struct TypeStruct3DerivedMixedParam {
	_a: TypeStruct1SimpleParam,
	_b: TypeStruct2DynamicParam,
	_c: TypeStruct2MixedParam,
}

#[test]
fn empty() {}

#[test]
fn impl_abi_type_signature() {
	assert_eq!(
		<TypeStruct1SimpleParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		"(uint8)"
	);
	assert_eq!(
		<TypeStruct1DynamicParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		"(string)"
	);
	assert_eq!(
		<TypeStruct2SimpleParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		"(uint8,uint32)"
	);
	assert_eq!(
		<TypeStruct2DynamicParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		"(string,bytes)"
	);
	assert_eq!(
		<TypeStruct2MixedParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		"(uint8,bytes)"
	);
	assert_eq!(
		<TypeStruct1DerivedSimpleParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		"((uint8))"
	);
	assert_eq!(
		<TypeStruct2DerivedSimpleParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		"((uint8),(uint8,uint32))"
	);
	assert_eq!(
		<TypeStruct1DerivedDynamicParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		"((string))"
	);
	assert_eq!(
		<TypeStruct2DerivedDynamicParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		"((string),(string,bytes))"
	);
	assert_eq!(
		<TypeStruct3DerivedMixedParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		"((uint8),(string,bytes),(uint8,bytes))"
	);
}

#[test]
fn impl_abi_type_is_dynamic() {
	assert_eq!(
		<TypeStruct1SimpleParam as evm_coder::abi::AbiType>::is_dynamic(),
		false
	);
	assert_eq!(
		<TypeStruct1DynamicParam as evm_coder::abi::AbiType>::is_dynamic(),
		true
	);
	assert_eq!(
		<TypeStruct2SimpleParam as evm_coder::abi::AbiType>::is_dynamic(),
		false
	);
	assert_eq!(
		<TypeStruct2DynamicParam as evm_coder::abi::AbiType>::is_dynamic(),
		true
	);
	assert_eq!(
		<TypeStruct2MixedParam as evm_coder::abi::AbiType>::is_dynamic(),
		true
	);
	assert_eq!(
		<TypeStruct1DerivedSimpleParam as evm_coder::abi::AbiType>::is_dynamic(),
		false
	);
	assert_eq!(
		<TypeStruct2DerivedSimpleParam as evm_coder::abi::AbiType>::is_dynamic(),
		false
	);
	assert_eq!(
		<TypeStruct1DerivedDynamicParam as evm_coder::abi::AbiType>::is_dynamic(),
		true
	);
	assert_eq!(
		<TypeStruct2DerivedDynamicParam as evm_coder::abi::AbiType>::is_dynamic(),
		true
	);
	assert_eq!(
		<TypeStruct3DerivedMixedParam as evm_coder::abi::AbiType>::is_dynamic(),
		true
	);
}

#[test]
fn impl_abi_type_size() {
	const ABI_ALIGNMENT: usize = 32;
	assert_eq!(
		<TypeStruct1SimpleParam as evm_coder::abi::AbiType>::size(),
		ABI_ALIGNMENT
	);
	assert_eq!(
		<TypeStruct1DynamicParam as evm_coder::abi::AbiType>::size(),
		ABI_ALIGNMENT
	);
	assert_eq!(
		<TypeStruct2SimpleParam as evm_coder::abi::AbiType>::size(),
		ABI_ALIGNMENT * 2
	);
	assert_eq!(
		<TypeStruct2DynamicParam as evm_coder::abi::AbiType>::size(),
		ABI_ALIGNMENT * 2
	);
	assert_eq!(
		<TypeStruct2MixedParam as evm_coder::abi::AbiType>::size(),
		ABI_ALIGNMENT * 2
	);
	assert_eq!(
		<TypeStruct1DerivedSimpleParam as evm_coder::abi::AbiType>::size(),
		ABI_ALIGNMENT
	);
	assert_eq!(
		<TypeStruct2DerivedSimpleParam as evm_coder::abi::AbiType>::size(),
		ABI_ALIGNMENT * 3
	);
	assert_eq!(
		<TypeStruct1DerivedDynamicParam as evm_coder::abi::AbiType>::size(),
		ABI_ALIGNMENT
	);
	assert_eq!(
		<TypeStruct2DerivedDynamicParam as evm_coder::abi::AbiType>::size(),
		ABI_ALIGNMENT * 3
	);
	assert_eq!(
		<TypeStruct3DerivedMixedParam as evm_coder::abi::AbiType>::size(),
		ABI_ALIGNMENT * 5
	);
}

#[derive(AbiCoder, PartialEq, Debug)]
struct TupleStruct1SimpleParam(u8);

#[derive(AbiCoder, PartialEq, Debug)]
struct TupleStruct1DynamicParam(String);

#[derive(AbiCoder, PartialEq, Debug)]
struct TupleStruct2SimpleParam(u8, u32);

#[derive(AbiCoder, PartialEq, Debug)]
struct TupleStruct2DynamicParam(String, bytes);

#[derive(AbiCoder, PartialEq, Debug)]
struct TupleStruct2MixedParam(u8, bytes);

#[derive(AbiCoder, PartialEq, Debug)]
struct TupleStruct1DerivedSimpleParam(TupleStruct1SimpleParam);

#[derive(AbiCoder, PartialEq, Debug)]
struct TupleStruct2DerivedSimpleParam(TupleStruct1SimpleParam, TupleStruct2SimpleParam);

#[derive(AbiCoder, PartialEq, Debug)]
struct TupleStruct1DerivedDynamicParam(TupleStruct1DynamicParam);

#[derive(AbiCoder, PartialEq, Debug)]
struct TupleStruct2DerivedDynamicParam(TupleStruct1DynamicParam, TupleStruct2DynamicParam);

#[derive(AbiCoder, PartialEq, Debug)]
struct TupleStruct3DerivedMixedParam(
	TupleStruct1SimpleParam,
	TupleStruct2DynamicParam,
	TupleStruct2MixedParam,
);

#[test]
fn impl_abi_type_signature_same_for_structs() {
	assert_eq!(
		<TypeStruct1SimpleParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		<TupleStruct1SimpleParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap()
	);
	assert_eq!(
		<TypeStruct1DynamicParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		<TupleStruct1DynamicParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap()
	);
	assert_eq!(
		<TypeStruct2SimpleParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		<TupleStruct2SimpleParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap()
	);
	assert_eq!(
		<TypeStruct2DynamicParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		<TupleStruct2DynamicParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap()
	);
	assert_eq!(
		<TypeStruct2MixedParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		<TupleStruct2MixedParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
	);
	assert_eq!(
		<TypeStruct1DerivedSimpleParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		<TupleStruct1DerivedSimpleParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
	);
	assert_eq!(
		<TypeStruct2DerivedSimpleParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		<TupleStruct2DerivedSimpleParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
	);
	assert_eq!(
		<TypeStruct1DerivedDynamicParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		<TupleStruct1DerivedDynamicParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
	);
	assert_eq!(
		<TypeStruct2DerivedDynamicParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		<TupleStruct2DerivedDynamicParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
	);
	assert_eq!(
		<TypeStruct3DerivedMixedParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
		<TupleStruct3DerivedMixedParam as evm_coder::abi::AbiType>::SIGNATURE
			.as_str()
			.unwrap(),
	);
}

#[test]
fn impl_abi_type_is_dynamic_same_for_structs() {
	assert_eq!(
		<TypeStruct1SimpleParam as evm_coder::abi::AbiType>::is_dynamic(),
		<TupleStruct1SimpleParam as evm_coder::abi::AbiType>::is_dynamic()
	);
	assert_eq!(
		<TypeStruct1DynamicParam as evm_coder::abi::AbiType>::is_dynamic(),
		<TupleStruct1DynamicParam as evm_coder::abi::AbiType>::is_dynamic()
	);
	assert_eq!(
		<TypeStruct2SimpleParam as evm_coder::abi::AbiType>::is_dynamic(),
		<TupleStruct2SimpleParam as evm_coder::abi::AbiType>::is_dynamic()
	);
	assert_eq!(
		<TypeStruct2DynamicParam as evm_coder::abi::AbiType>::is_dynamic(),
		<TupleStruct2DynamicParam as evm_coder::abi::AbiType>::is_dynamic()
	);
	assert_eq!(
		<TypeStruct2MixedParam as evm_coder::abi::AbiType>::is_dynamic(),
		<TupleStruct2MixedParam as evm_coder::abi::AbiType>::is_dynamic()
	);
	assert_eq!(
		<TypeStruct1DerivedSimpleParam as evm_coder::abi::AbiType>::is_dynamic(),
		<TupleStruct1DerivedSimpleParam as evm_coder::abi::AbiType>::is_dynamic()
	);
	assert_eq!(
		<TypeStruct2DerivedSimpleParam as evm_coder::abi::AbiType>::is_dynamic(),
		<TupleStruct2DerivedSimpleParam as evm_coder::abi::AbiType>::is_dynamic()
	);
	assert_eq!(
		<TypeStruct1DerivedDynamicParam as evm_coder::abi::AbiType>::is_dynamic(),
		<TupleStruct1DerivedDynamicParam as evm_coder::abi::AbiType>::is_dynamic()
	);
	assert_eq!(
		<TypeStruct2DerivedDynamicParam as evm_coder::abi::AbiType>::is_dynamic(),
		<TupleStruct2DerivedDynamicParam as evm_coder::abi::AbiType>::is_dynamic()
	);
	assert_eq!(
		<TypeStruct3DerivedMixedParam as evm_coder::abi::AbiType>::is_dynamic(),
		<TupleStruct3DerivedMixedParam as evm_coder::abi::AbiType>::is_dynamic()
	);
}

#[test]
fn impl_abi_type_size_same_for_structs() {
	assert_eq!(
		<TypeStruct1SimpleParam as evm_coder::abi::AbiType>::size(),
		<TupleStruct1SimpleParam as evm_coder::abi::AbiType>::size()
	);
	assert_eq!(
		<TypeStruct1DynamicParam as evm_coder::abi::AbiType>::size(),
		<TupleStruct1DynamicParam as evm_coder::abi::AbiType>::size()
	);
	assert_eq!(
		<TypeStruct2SimpleParam as evm_coder::abi::AbiType>::size(),
		<TupleStruct2SimpleParam as evm_coder::abi::AbiType>::size()
	);
	assert_eq!(
		<TypeStruct2DynamicParam as evm_coder::abi::AbiType>::size(),
		<TupleStruct2DynamicParam as evm_coder::abi::AbiType>::size()
	);
	assert_eq!(
		<TypeStruct2MixedParam as evm_coder::abi::AbiType>::size(),
		<TupleStruct2MixedParam as evm_coder::abi::AbiType>::size()
	);
	assert_eq!(
		<TypeStruct1DerivedSimpleParam as evm_coder::abi::AbiType>::size(),
		<TupleStruct1DerivedSimpleParam as evm_coder::abi::AbiType>::size()
	);
	assert_eq!(
		<TypeStruct2DerivedSimpleParam as evm_coder::abi::AbiType>::size(),
		<TupleStruct2DerivedSimpleParam as evm_coder::abi::AbiType>::size()
	);
	assert_eq!(
		<TypeStruct1DerivedDynamicParam as evm_coder::abi::AbiType>::size(),
		<TupleStruct1DerivedDynamicParam as evm_coder::abi::AbiType>::size()
	);
	assert_eq!(
		<TypeStruct2DerivedDynamicParam as evm_coder::abi::AbiType>::size(),
		<TupleStruct2DerivedDynamicParam as evm_coder::abi::AbiType>::size()
	);
	assert_eq!(
		<TypeStruct3DerivedMixedParam as evm_coder::abi::AbiType>::size(),
		<TupleStruct3DerivedMixedParam as evm_coder::abi::AbiType>::size()
	);
}

const FUNCTION_IDENTIFIER: u32 = 0xdeadbeef;

fn test_impl<Tuple, TupleStruct, TypeStruct>(
	tuple_data: Tuple,
	tuple_struct_data: TupleStruct,
	type_struct_data: TypeStruct,
) where
	TypeStruct:
		evm_coder::abi::AbiWrite + evm_coder::abi::AbiRead + std::cmp::PartialEq + std::fmt::Debug,
	TupleStruct:
		evm_coder::abi::AbiWrite + evm_coder::abi::AbiRead + std::cmp::PartialEq + std::fmt::Debug,
	Tuple:
		evm_coder::abi::AbiWrite + evm_coder::abi::AbiRead + std::cmp::PartialEq + std::fmt::Debug,
{
	let encoded_type_struct = test_abi_write_impl(&type_struct_data);
	let encoded_tuple_struct = test_abi_write_impl(&tuple_struct_data);
	let encoded_tuple = test_abi_write_impl(&tuple_data);

	similar_asserts::assert_eq!(encoded_tuple, encoded_type_struct);
	similar_asserts::assert_eq!(encoded_tuple, encoded_tuple_struct);

	{
		let (_, mut decoder) = evm_coder::abi::AbiReader::new_call(&encoded_tuple).unwrap();
		let restored_struct_data = <TypeStruct>::abi_read(&mut decoder).unwrap();
		assert_eq!(restored_struct_data, type_struct_data);
	}
	{
		let (_, mut decoder) = evm_coder::abi::AbiReader::new_call(&encoded_tuple).unwrap();
		let restored_struct_data = <TupleStruct>::abi_read(&mut decoder).unwrap();
		assert_eq!(restored_struct_data, tuple_struct_data);
	}

	{
		let (_, mut decoder) = evm_coder::abi::AbiReader::new_call(&encoded_type_struct).unwrap();
		let restored_tuple_data = <Tuple>::abi_read(&mut decoder).unwrap();
		assert_eq!(restored_tuple_data, tuple_data);
	}
	{
		let (_, mut decoder) = evm_coder::abi::AbiReader::new_call(&encoded_tuple_struct).unwrap();
		let restored_tuple_data = <Tuple>::abi_read(&mut decoder).unwrap();
		assert_eq!(restored_tuple_data, tuple_data);
	}
}

fn test_abi_write_impl<A>(data: &A) -> Vec<u8>
where
	A: evm_coder::abi::AbiWrite + evm_coder::abi::AbiRead + std::cmp::PartialEq + std::fmt::Debug,
{
	let mut writer = evm_coder::abi::AbiWriter::new_call(FUNCTION_IDENTIFIER);
	data.abi_write(&mut writer);
	let encoded_tuple = writer.finish();
	encoded_tuple
}

#[test]
fn codec_struct_1_simple() {
	let _a = 0xff;
	test_impl::<(u8,), TupleStruct1SimpleParam, TypeStruct1SimpleParam>(
		(_a,),
		TupleStruct1SimpleParam(_a),
		TypeStruct1SimpleParam { _a },
	);
}

#[test]
fn codec_struct_1_dynamic() {
	let _a: String = "some string".into();
	test_impl::<(String,), TupleStruct1DynamicParam, TypeStruct1DynamicParam>(
		(_a.clone(),),
		TupleStruct1DynamicParam(_a.clone()),
		TypeStruct1DynamicParam { _a },
	);
}

#[test]
fn codec_struct_1_derived_simple() {
	let _a: u8 = 0xff;
	test_impl::<((u8,),), TupleStruct1DerivedSimpleParam, TypeStruct1DerivedSimpleParam>(
		((_a,),),
		TupleStruct1DerivedSimpleParam(TupleStruct1SimpleParam(_a)),
		TypeStruct1DerivedSimpleParam {
			_a: TypeStruct1SimpleParam { _a },
		},
	);
}

#[test]
fn codec_struct_1_derived_dynamic() {
	let _a: String = "some string".into();
	test_impl::<((String,),), TupleStruct1DerivedDynamicParam, TypeStruct1DerivedDynamicParam>(
		((_a.clone(),),),
		TupleStruct1DerivedDynamicParam(TupleStruct1DynamicParam(_a.clone())),
		TypeStruct1DerivedDynamicParam {
			_a: TypeStruct1DynamicParam { _a },
		},
	);
}

#[test]
fn codec_struct_2_simple() {
	let _a = 0xff;
	let _b = 0xbeefbaba;
	test_impl::<(u8, u32), TupleStruct2SimpleParam, TypeStruct2SimpleParam>(
		(_a, _b),
		TupleStruct2SimpleParam(_a, _b),
		TypeStruct2SimpleParam { _a, _b },
	);
}

#[test]
fn codec_struct_2_dynamic() {
	let _a: String = "some string".into();
	let _b: bytes = bytes(vec![0x11, 0x22, 0x33]);
	test_impl::<(String, bytes), TupleStruct2DynamicParam, TypeStruct2DynamicParam>(
		(_a.clone(), _b.clone()),
		TupleStruct2DynamicParam(_a.clone(), _b.clone()),
		TypeStruct2DynamicParam { _a, _b },
	);
}

#[test]
fn codec_struct_2_mixed() {
	let _a: u8 = 0xff;
	let _b: bytes = bytes(vec![0x11, 0x22, 0x33]);
	test_impl::<(u8, bytes), TupleStruct2MixedParam, TypeStruct2MixedParam>(
		(_a.clone(), _b.clone()),
		TupleStruct2MixedParam(_a.clone(), _b.clone()),
		TypeStruct2MixedParam { _a, _b },
	);
}

#[test]
fn codec_struct_2_derived_simple() {
	let _a = 0xff;
	let _b = 0xbeefbaba;
	test_impl::<((u8,), (u8, u32)), TupleStruct2DerivedSimpleParam, TypeStruct2DerivedSimpleParam>(
		((_a,), (_a, _b)),
		TupleStruct2DerivedSimpleParam(
			TupleStruct1SimpleParam(_a),
			TupleStruct2SimpleParam(_a, _b),
		),
		TypeStruct2DerivedSimpleParam {
			_a: TypeStruct1SimpleParam { _a },
			_b: TypeStruct2SimpleParam { _a, _b },
		},
	);
}

#[test]
fn codec_struct_2_derived_dynamic() {
	let _a = "some string".to_string();
	let _b = bytes(vec![0x11, 0x22, 0x33]);
	test_impl::<
		((String,), (String, bytes)),
		TupleStruct2DerivedDynamicParam,
		TypeStruct2DerivedDynamicParam,
	>(
		((_a.clone(),), (_a.clone(), _b.clone())),
		TupleStruct2DerivedDynamicParam(
			TupleStruct1DynamicParam(_a.clone()),
			TupleStruct2DynamicParam(_a.clone(), _b.clone()),
		),
		TypeStruct2DerivedDynamicParam {
			_a: TypeStruct1DynamicParam { _a: _a.clone() },
			_b: TypeStruct2DynamicParam { _a, _b },
		},
	);
}

#[test]
fn codec_struct_3_derived_mixed() {
	let int = 0xff;
	let by = bytes(vec![0x11, 0x22, 0x33]);
	let string = "some string".to_string();
	test_impl::<
		((u8,), (String, bytes), (u8, bytes)),
		TupleStruct3DerivedMixedParam,
		TypeStruct3DerivedMixedParam,
	>(
		((int,), (string.clone(), by.clone()), (int, by.clone())),
		TupleStruct3DerivedMixedParam(
			TupleStruct1SimpleParam(int),
			TupleStruct2DynamicParam(string.clone(), by.clone()),
			TupleStruct2MixedParam(int, by.clone()),
		),
		TypeStruct3DerivedMixedParam {
			_a: TypeStruct1SimpleParam { _a: int },
			_b: TypeStruct2DynamicParam {
				_a: string.clone(),
				_b: by.clone(),
			},
			_c: TypeStruct2MixedParam { _a: int, _b: by },
		},
	);
}

#[derive(AbiCoder, PartialEq, Debug)]
struct TypeStruct2SimpleStruct1Simple {
	_a: TypeStruct2SimpleParam,
	_b: TypeStruct2SimpleParam,
	_c: u8,
}
#[derive(AbiCoder, PartialEq, Debug)]
struct TupleStruct2SimpleStruct1Simple(TupleStruct2SimpleParam, TupleStruct2SimpleParam, u8);

#[test]
fn codec_struct_2_struct_simple_1_simple() {
	let _a = 0xff;
	let _b = 0xbeefbaba;
	test_impl::<
		((u8, u32), (u8, u32), u8),
		TupleStruct2SimpleStruct1Simple,
		TypeStruct2SimpleStruct1Simple,
	>(
		((_a, _b), (_a, _b), _a),
		TupleStruct2SimpleStruct1Simple(
			TupleStruct2SimpleParam(_a, _b),
			TupleStruct2SimpleParam(_a, _b),
			_a,
		),
		TypeStruct2SimpleStruct1Simple {
			_a: TypeStruct2SimpleParam { _a, _b },
			_b: TypeStruct2SimpleParam { _a, _b },
			_c: _a,
		},
	);
}
