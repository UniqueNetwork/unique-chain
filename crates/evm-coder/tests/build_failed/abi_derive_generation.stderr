error: Empty structs not supported
 --> tests/build_failed/abi_derive_generation.rs:4:8
  |
4 | struct EmptyStruct {}
  |        ^^^^^^^^^^^

error: Empty structs not supported
 --> tests/build_failed/abi_derive_generation.rs:7:8
  |
7 | struct EmptyTupleStruct();
  |        ^^^^^^^^^^^^^^^^
